src/rust/engine/process_execution/src/workspace.rs

```diff
@@ -0,0 +1,262 @@
+
+        let exclusive_spawn = prepare_workdir(
+            tempdir.path().to_owned(),
+            &self.work_dir_base,
+            &req,
+            req.input_digests.inputs.clone(),
+            &self.store,
+            &self.named_caches,
+            &self.immutable_inputs,
+            None,
+            None,
+        )
+        .await?;
+
+        apply_chroot(tempdir.path().to_str().unwrap(), &mut req);
+
+        let res = self
+            .run_and_capture_workdir(
+                req.clone(),
+                context,
+                self.store.clone(),
+                self.executor.clone(),
+                tempdir.path().to_owned(),
+                self.build_root.clone(),
+                exclusive_spawn,
+            )
+            .map_err(|msg| {
+                // Processes that experience no infrastructure issues should result in an "Ok" return,
+                // potentially with an exit code that indicates that they failed (with more information
+                // on stderr). Actually failing at this level indicates a failure to start or otherwise
+                // interact with the process, which would generally be an infrastructure or implementation
+                // error (something missing from the sandbox, incorrect permissions, etc).
+                //
+                // Given that this is expected to be rare, we dump the entire process definition in the
+                // error.
+                ProcessError::Unclassified(format!("Failed to execute: {req_debug_repr}\n\n{msg}"))
+            })
+            .await;
+
+        res
+    }
+
+    async fn shutdown(&self) -> Result<(), String> {
+        Ok(())
+    }
+}
+
+impl fmt::Debug for CommandRunner {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("workspace::CommandRunner")
+            .finish_non_exhaustive()
+    }
+}
+
+#[async_trait]
+impl CapturedWorkdir for CommandRunner {
+    type WorkdirToken = PathBuf;
+
+    fn apply_working_directory_to_outputs() -> bool {
+        false
+    }
+
+    async fn run_in_workdir<'s, 'c, 'w, 'r>(
+        &'s self,
+        _context: &'c Context,
+        _workdir_path: &'w Path,
+        build_root: Self::WorkdirToken,
+        req: Process,
+        exclusive_spawn: bool,
+    ) -> Result<BoxStream<'r, Result<ChildOutput, String>>, String> {
+        let cwd = if let Some(working_directory) = &req.working_directory {
+            build_root.join(working_directory)
+        } else {
+            build_root
+        };
+        let mut command = Command::new(&req.argv[0]);
+        command
+            .env_clear()
+            .args(&req.argv[1..])
+            .current_dir(cwd)
+            .envs(&req.env)
+            .stdin(Stdio::null())
+            .stdout(Stdio::piped())
+            .stderr(Stdio::piped());
+
+        // See the documentation of the `CapturedWorkdir::run_in_workdir` method, but `exclusive_spawn`
```

Comment:
This section is duplicated with `local::CommandRunner` and should be refactored.

Reply:
Done.